controller/engine_controller.go

```diff
@@ -501,15 +501,42 @@ func (ec *EngineController) DeleteInstance(obj interface{}) error {
 		}
 	}
 
-	if im.Status.CurrentState != longhorn.InstanceManagerStateRunning {
-		return nil
+	v, err := ec.ds.GetVolumeRO(e.Spec.VolumeName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	isRWXVolume := false
+	if v != nil && v.Spec.AccessMode == longhorn.AccessModeReadWriteMany {
+		isRWXVolume = true
+	}
+
+	if !isRWXVolume {
+		if im.Status.CurrentState != longhorn.InstanceManagerStateRunning {
+			logrus.Infof("Skip deleting engine %v for volume %v since instance manager is in %v state", e.Name, v.Name, im.Status.CurrentState)
+			return nil
+		}
 	}
 
+	logrus.Infof("Deleting engine %v for volume %v", e.Name, v.Name)
+
+	defer func() {
+		logrus.WithError(err).Warnf("Failed to delete engine %v for volume %v", e.Name, v.Name)
+		if isRWXVolume && im.Status.CurrentState != longhorn.InstanceManagerStateRunning {
+			// Try the best to delete engine and ignore the error when volume is
+			// a RWX volume and the instance manager is not running
+			err = nil
+		}
+	}()
```

Comment:
Why we need this? I assume we only need to allow the engine process is able to delete for the volume used by the share manager pod. 

If somehow it can't be deleted, this controller should periodically reconcile this to retry the deletion. It seems no need to ignore the returned error, but still respect any error happening here. WDYT?

BTW, `err` is a local variable, so if need to deal with this way, probably should use named returned variable?

Reply:
Yeah, I think reconciling the deletion is better.